### PR TITLE
:sparkles: summarize best practices for data science session

### PR DESCRIPTION
**Pull request by enryH**
Based on session on internal retreat on the 18th of September 2024.

### CODE REVIEW

File: python/best_practices.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+```toml 
+# percent format and ipynb format in sync
+formats = "ipynb,py:percent"
+```
+
+## Copilot in VSCode
+
+Ghosttext, chats and inline chats are great ways to get suggestions on the code you are 
+writing. You can apply for a free version as a [(PhD) student](https://github.com/education/students)
+or [instructor](https://github.com/education/teachers). Currently alternatives wit a free-tier as [codium](https://codeium.com/) are also available.
```
Comment (JakobBerg) [inline python/best_practices.md:71]: *with

File: python/best_practices.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Coding (best) practices for Data Science
+
+> Author: Henry Webel
+
+Being asked to show some coding best practices for an internal retreat, I assembeled
```
Comment (pcolaianni) [inline python/best_practices.md:5]:
```suggestion
Being asked to show some coding best practices for an internal retreat, I assembled
```

File: python/best_practices.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+some low hanging fruits in reach for everyone and some pratices I learned to appreciate.
```
Comment (pcolaianni) [inline python/best_practices.md:6]:
```suggestion
some low hanging fruits in reach for everyone and some practices I learned to appreciate.
```

File: python/best_practices.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+## Use a formater
```
Comment (pcolaianni) [inline python/best_practices.md:8]:
```suggestion
## Use a formatter
```

File: python/best_practices.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+When you write code, you should at least use some sort of formatter. `black` is common choice
```
Comment (pcolaianni) [inline python/best_practices.md:10]:
```suggestion
When you write code, I encourage using a formatter. `black` is a common choice
```

File: python/best_practices.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+as it allows you to format code in a user-defined linelength consistently. It even can
+break too long strings into it's parts - leaving only long comments and docstrings to you
```
Comment (pcolaianni) [inline python/best_practices.md:12]:
```suggestion
break too long strings into its parts - leaving only long comments and docstrings to you
```

File: python/best_practices.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+for adoption.
+
+`black` or `autopep8` are also availbe next to `isort` for sorting imports in VSCode as
```
Comment (pcolaianni) [inline python/best_practices.md:15]:
```suggestion
`black` or `autopep8` are also available next to `isort` for sorting imports in VSCode as 
```

File: python/best_practices.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+`black` or `autopep8` are also availbe next to `isort` for sorting imports in VSCode as 
+extension, so your files are formatted everytime you save these 
+([link](https://code.visualstudio.com/docs/python/formatting)).
+
+## Use a linter
+
+Too long lines, unpassed arguments or mutable objects as default function parameters you can
+identify using a linter like `flake8` or `ruff`. Tools like 
+[`Pylint` in VSCode](https://code.visualstudio.com/docs/python/linting)
+allow you to get in editor highlighting of Code issues and links with hints on how to fix them
```
Comment (pcolaianni) [inline python/best_practices.md:24]:
```suggestion
allow your code editor to highlight issues and links with hints on how to fix them
```

File: python/best_practices.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+for adoption.
+
+`black` or `autopep8` are also availbe next to `isort` for sorting imports in VSCode as 
+extension, so your files are formatted everytime you save these 
+([link](https://code.visualstudio.com/docs/python/formatting)).
+
+## Use a linter
+
+Too long lines, unpassed arguments or mutable objects as default function parameters you can
+identify using a linter like `flake8` or `ruff`. Tools like
```
Comment (pcolaianni) [inline python/best_practices.md:22]: I would invert the two sentences: Using a linter blabla helps you identifying blabla

File: python/best_practices.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+## Use a linter
+
+Too long lines, unpassed arguments or mutable objects as default function parameters you can
+identify using a linter like `flake8` or `ruff`. Tools like 
+[`Pylint` in VSCode](https://code.visualstudio.com/docs/python/linting)
+allow you to get in editor highlighting of Code issues and links with hints on how to fix them
+![screenshot with typehints](assets/lint_pylance_in_vscode.png)
+
+Example: Using the linter you can for example if you did not pass an argument to a function
```
Comment (pcolaianni) [inline python/best_practices.md:27]: There is a word or two missing here. And repetition of "example".

File: python/best_practices.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+[ToDo Tree](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) 
+you can easily keep a list of todos in your code - allowing you to go 
+through them from time to time and prioritze.
+
+![Highlighted Comments and ToDo Tree example](assets/better_comments_todo_tree.png)
+
+## Text based Notebook (percent format) with jupytext and papermill
+
+[`jupytext`](https://jupytext.readthedocs.io/) is a lightweight tool to keep scripts either as notebooks (`.ipynb`) or simpler text based file formats, such as markdown files (`.md`)  which can be easily rendered on GitHub or python files  (`.py`) which can be executed in VSCode’s interactive shell and are better for version control. Some tools still need ipynb to work, e.g. `papermill`. Therefore it is handy to keep different version of a script in sync. Otherwise one can also only use python files and render these as notebook in e.g. 
+[jupyter lab](https://jupytext.readthedocs.io/en/latest/text-notebooks.html). Especially if the code is only kept for version control, but executed versions are keep in a project folder using a workflow environment (as `snakemake` or `nextflow`) this comes in handy.
```
Comment (pcolaianni) [inline python/best_practices.md:46]:
```suggestion
[jupyter lab](https://jupytext.readthedocs.io/en/latest/text-notebooks.html). This is useful especially if the code is only kept for version control, but executed versions are kept in a project folder using a workflow environment (as `snakemake` or `nextflow`).
```

**Comment by enryH** [comment]
Thanks for the corrections!

**Pull request merged**